keras_nlp/models/xlnet/xlnet_tokenizer.py

```diff
@@ -0,0 +1,205 @@
+
+    Examples:
+
+    ```python
+    # Unbatched input.
+    tokenizer = keras_nlp.models.XLNetTokenizer(proto="<path to SentencePiece proto file>")
+    tokenizer("The quick brown fox jumped.")
+
+    # Batched input.
+    tokenizer(["The quick brown fox jumped.", "The fox slept."])
+
+    # Detokenization.
+    tokenizer.detokenize(tokenizer("The quick brown fox jumped."))
+
+    # Custom vocabulary.
+    bytes_io = io.BytesIO()
+    ds = tf.data.Dataset.from_tensor_slices(["The quick brown fox jumped."])
+    sentencepiece.SentencePieceTrainer.train(
+        sentence_iterator=vocab_data.as_numpy_iterator(),
+        model_writer=bytes_io,
+        vocab_size=14,
+        model_type="WORD",
+        pad_id=0,
+        bos_id=1,
+        eos_id=2,
+        unk_id=3,
+        pad_piece="<pad>",
+        bos_piece="<s>",
+        eos_piece="</s>",
+        unk_piece="<unk>",
+        user_defined_symbols=["<mask>", "<cls>", "<sep>"]
+    )
+    tokenizer = keras_nlp.models.XLNetTokenizer(
+        proto=bytes_io.getvalue(),
+    )
+    tokenizer("The quick brown fox jumped.")
+    ```
+    """
+
+    def __init__(self, proto, **kwargs):
+        super().__init__(proto=proto, **kwargs)
+
+        # Check for necessary special tokens.
+        self.cls_token = "<cls>"
+        self.sep_token = "<sep>"
+        self.pad_token = "<pad>"
+        self.mask_token = "<mask>"
+        self.bos_token = "<s>"
+        self.eos_token = "</s>"
+        self.unk_token = "<unk>"
+
+        for token in [
+            self.cls_token,
+            self.sep_token,
+            self.pad_token,
+            self.mask_token,
+            self.bos_token,
+            self.eos_token,
+            self.unk_token,
+        ]:
+            if token not in self.get_vocabulary():
+                raise ValueError(
+                    f"Cannot find token `'{token}'` in the provided "
+                    f"`vocabulary`. Please provide `'{token}'` in your "
+                    "`vocabulary` or use a pretrained `vocabulary` name."
+                )
+
+        self.cls_token_id = self.token_to_id(self.cls_token)
+        self.sep_token_id = self.token_to_id(self.sep_token)
+        self.pad_token_id = self.token_to_id(self.pad_token)
+        self.mask_token_id = self.token_to_id(self.mask_token)
+        self.bos_token_id = self.token_to_id(self.bos_token)
+        self.eos_token_id = self.token_to_id(self.eos_token)
+        self.unk_token_id = self.token_to_id(self.unk_token)
+
+    def preprocess_text(self, inputs):
+        """Preprocesses the text. This method removes spaces and accents from the text."""
+
+        # remove space
+        outputs = " ".join(inputs.strip().split())
+        outputs = outputs.replace("``", '"').replace("''", '"')
+
+        # remove accents
+        outputs = unicodedata.normalize("NFKD", outputs)
+        outputs = "".join([c for c in outputs if not unicodedata.combining(c)])
+
+        return outputs
+
+    def tokenize(self, text):
```

Comment:
This does not look like it would work with `tf.data`. A key feature for our tokenizers is to be able to run `string_ds.map(tokenizer)`, with a `tf.data.Dataset`, as this is really the only performant option for preprocessing we ship with the library.

I would not worry about being one to one with huggingface w.r.t. string inputted special tokens right now, but we do need two things...
- Plain text (ignore special tokens in both input and output), should tokenize exactly the same as the upstream implementation.
- `tokenize()` should chain to `super()` and the `tf.text` op for tokenizing text. No for loop tokenization.

If we can get to that state we will be unblocked here. Why is there a need to diverge from the sentence piece routines below?

Reply:
Hi @mattdangerw thanks for you comment! Yes the tokenizer is not working with the `tf.data.Dataset`. 

>Plain text (ignore special tokens in both input and output), should tokenize exactly the same as the upstream implementation.

For (almost) all plain texts the `super().tokenize` is enough and produces the same upstream result but there are a few texts (such as `"ABC 0.123,"`) where we must apply the extra logic to get the same result. 

- Output of `[tokenizer.id_to_token(i) for i in tokenizer._sentence_piece.tokenize("ABC 0.123,")]` - 
`['▁ABC', '▁0', '.', '12', '3,']`

But the actual output is `['▁ABC', '▁0', '.', '12', '3', ',']`
So, we must keep the extra logic in the tokenize. (The official repo also has the [same logic](https://github.com/zihangdai/xlnet/blob/bbaa3a6fa0b3a2ee694e8cf66167434f9eca9660/prepro_utils.py#L80))

My current plan is to replace all other `str` methods with `tf text` and remove the outer loop.

Reply:
Thanks for the explainer! Is it really just some weird workaround for digits followed by a comma?

Ideally we could figure out a way to either preprocess or postprocess the sentencepiece tokenize result so that we can still use the `tf-text` sentencepiece "black box" unaltered. Not sure if that is possible though...

`tensorflow-text` and the `tf.strings` module will be a main tools here.

Reply:
Yes as of my understanding it's a workaround.